### PR TITLE
Reduce AssemblyName allocations

### PR DESCRIPTION
**Pull request by cdmihai**
In mvc rebuild (133mb total allocations), reduces AssemblyName allocations 18.3% ->  0.5% and AssemblyNameExtensions 4.5% -> 0.3%. In a Roslyn rebuild (3.2gb total allocations), it reduces AssemblyName 2.3% -> 0.2% and AsssemblyNameExtension 0.6% -> 0.1%.

Commit [#1](https://github.com/Microsoft/msbuild/commit/8d5b139b9fabe58d288a65d4c34fe887fe812ae5) is safe to do, because the AssemblyName was double cloned ...
Commit [#2](https://github.com/Microsoft/msbuild/pull/2676/commits/b292563136731a53858defbac2fda9b1d056d7af) replaces calls to [AssemblyNameExtension.EqualsImpl](https://github.com/Microsoft/msbuild/blob/0a47f62a138b32cb406d058fe7f49c0494aa6c84/src/Shared/AssemblyNameExtension.cs#L670-L733) and [AssemblyNameExtension.CompareBaseNameToImpl](https://github.com/Microsoft/msbuild/blob/0a47f62a138b32cb406d058fe7f49c0494aa6c84/src/Shared/AssemblyNameExtension.cs#L516-L546) with a simple name, culture, and key comparison. This _looks_ safe to do because:
- the versions don't matter because both assemblies [get the same version](https://github.com/Microsoft/msbuild/blob/0a47f62a138b32cb406d058fe7f49c0494aa6c84/src/Tasks/AssemblyDependency/ReferenceTable.cs#L2379)
- [AssemblyNameExtension.asString](https://github.com/Microsoft/msbuild/blob/0a47f62a138b32cb406d058fe7f49c0494aa6c84/src/Shared/AssemblyNameExtension.cs#L63) does not matter because the comparison assembly [is instantiated](https://github.com/Microsoft/msbuild/blob/0a47f62a138b32cb406d058fe7f49c0494aa6c84/src/Tasks/AssemblyDependency/ReferenceTable.cs#L2375) via a constructor that leaves `asString` null
- [ignoreVersion ](https://github.com/Microsoft/msbuild/blob/0a47f62a138b32cb406d058fe7f49c0494aa6c84/src/Shared/AssemblyNameExtension.cs#L712)and [considerRetargetableFlag ](https://github.com/Microsoft/msbuild/blob/0a47f62a138b32cb406d058fe7f49c0494aa6c84/src/Shared/AssemblyNameExtension.cs#L727) are [false](https://github.com/Microsoft/msbuild/blob/0a47f62a138b32cb406d058fe7f49c0494aa6c84/src/Shared/AssemblyNameExtension.cs#L646)

### CODE REVIEW

File: src/Shared/AssemblyNameExtension.cs [hunk 1]
```diff
@@ -734,29 +734,13 @@ private bool EqualsImpl(AssemblyNameExtension that, bool ignoreVersion, bool con
...
-                bCulture = CultureInfo.InvariantCulture;
-            }
-
-            if (aCulture.LCID != bCulture.LCID)
-            {
-                return false;
-            }
-
-            return true;
+            return a?.CultureInfo?.LCID == b?.CultureInfo?.LCID;
```
Comment (davkean) [inline src/Shared/AssemblyNameExtension.cs]: This isn't the same comparison as before - if one is Invariant and one is null, this used to return true, now it returns false.

Comment (cdmihai) [reply src/Shared/AssemblyNameExtension.cs]: You're right. I made the .net core implementation follow the full framework one and not the other way around

**Comment by cdmihai** [comment]
Apparently `BuildEnvironmentHelper.Instance.RunningTests` does not work on .net core :(. Looking into it.

File: src/Build.UnitTests/BuildEnvironmentHelper_Tests.cs [hunk 2]
```diff
@@ -19,6 +20,13 @@ public class BuildEnvironmentHelper_Tests
 #else
         private const string MSBuildExeName = "MSBuild.exe";
 #endif
+
+        [Fact]
+        public void ShouldBeRunningInTests()
```
Comment (AndyGerlicher) [inline src/Build.UnitTests/BuildEnvironmentHelper_Tests.cs]: https://github.com/cdmihai/msbuild/blob/ca62c0d4a3d9ba003176c4069a976978c59949d7/src/Build.UnitTests/BuildEnvironmentHelper_Tests.cs#L275-L284

😄

Comment (cdmihai) [reply src/Build.UnitTests/BuildEnvironmentHelper_Tests.cs]: Best kind of bugfix. :)

**Comment by davkean** [comment]
Can we get this merged in?

**Comment by AndyGerlicher** [comment]
Will submit this plus the others on Monday for escrow.

**Pull request merged**